go-controller/pkg/node/controllers/upgrade/node_upgrade.go

```diff
@@ -6,231 +6,133 @@ import (
+	return wait.PollWithContext(ctx, 10*time.Second, timeout, func(ctx context.Context) (bool, error) {
+		ver, err := uc.getVersion(ctx)
+		if err == nil {
+			if ver >= minVersion {
+				klog.Infof("Cluster topology version is now %d", ver)
+				return true, nil
+			}
 
-	done, err := uc.detectUpgradeDone(eKey.(string))
-	if err != nil {
-		utilruntime.HandleError(err)
-	}
-	uc.queue.Forget(eKey)
-	return done
+			klog.Infof("Cluster topology version %d < %d", ver, minVersion)
+			return false, nil
+		}
+		klog.Errorf("Failed to retrieve topology version: %v", err)
+		return false, nil // swallow error so we retry
+	})
 }
 
-// returns true if upgrade is detected as complete
-func (uc *upgradeController) detectUpgradeDone(nodeName string) (bool, error) {
-	klog.Infof("Processing node update for %s", nodeName)
-	// Get current Node from the cache
-	node, err := uc.nodeLister.Get(nodeName)
-	// It´s unlikely that we have an error different that "Not Found Object"
-	// because we are getting the object from the informer´s cache
+// getVersion retrieves the current version
+func (uc *upgradeController) getVersion(ctx context.Context) (int, error) {
+	// First, check the config map
+	ns := globalconfig.Kubernetes.OVNConfigNamespace
+	name := ovntypes.OvnK8sStatusCMName
+	cm, err := uc.client.CoreV1().ConfigMaps(ns).Get(ctx, name, metav1.GetOptions{})
 	if err != nil && !apierrors.IsNotFound(err) {
-		return false, err
+		klog.Warningf("Error retrieving ConfigMap %s/%s: %v", ns, name, err)
+		return -1, err
 	}
-	topoVer, ok := node.Annotations[ovntypes.OvnK8sTopoAnno]
-	if ok && len(topoVer) > 0 {
-		ver, err := strconv.Atoi(topoVer)
-		if err != nil {
-			klog.Warningf("Illegal value detected for %s, on node: %s, value: %s, error: %v",
-				ovntypes.OvnK8sTopoAnno, node.Name, topoVer, err)
-			return false, err
-		}
-		if ver == ovntypes.OvnCurrentTopologyVersion {
-			return true, nil
+	if err == nil {
+		tv, ok := cm.Data[ovntypes.OvnK8sStatusKeyTopoVersion]
+		if ok {
+			out, err := strconv.Atoi(tv)
+			if err == nil {
+				klog.Infof("Detected cluster topology version %d from ConfigMap %s/%s", out, ns, name)
+				return out, nil
+			} else {
+				klog.Infof("ConfigMap %s/%s had invalid value in %s: %s %v", ns, name, ovntypes.OvnK8sStatusKeyTopoVersion, tv, err)
+			}
 		}
 	}
-	return false, nil
-}
 
-// DetermineOVNTopoVersionOnNode determines what OVN Topology version is being used
-func (uc *upgradeController) detectInitialVersion(selector labels.Selector) (int, error) {
-	// Find out the current OVN topology version by checking "k8s.ovn.org/topo-version" annotation on Master nodes
-	nodeList, err := uc.client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{
-		LabelSelector: selector.String(),
-	})
+	// Masters not yet upgraded, check Node objects
+	klog.Infof("Could not determine TopologyVersion via configmap, falling back to Nodes")
+
+	nodes, err := uc.wf.GetNodes()
```

Comment:
you are missing the label selector to only get master nodes

Reply:
Because we end up looping over the nodes. So we should strive to write code that that will loop 3 times instead of 500.

Reply:
ok, added the label selector.